src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,348 @@
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 6) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test indexing draw with index format of uint32. If this is interpreted as uint16, then it would
+// be 0, 1, 0, ... and would draw nothing. And the index buffer size - offset must be not less than
+// the size required by triangle list, otherwise it also render nothing.
+g.test('index_format_uint32')
+  .params(params().combine(poptions('indexOffset', [0, 12, 36])))
+  .fn(t => {
+    const { indexOffset } = t.params;
+
+    const indexArray = new Uint32Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint32');
+    const indexCount = indexArray.length;
+    t.run(indexBuffer, indexCount, 'uint32', indexOffset);
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 12) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test primitive restart with each primitive topology. The primitive restart value can used with
+// strip primitive topologies ('line-strip' or 'triangle-strip').
+// For line-strip, [0, 1, PRIM_RESTART, 2, 3] only render a X shape.
+// For triangle-strip, [0, 1, 3, PRIM_RESTART, 2] only the first triangle.
+// Others are expected to be renderred the same as without primitive restart.
+g.test('primitive_restart')
+  .params(
+    params()
+      .combine(poptions('indexFormat', ['uint16', 'uint32']))
+      .combine(
+        poptions('primitiveTopology', [
+          'point-list',
+          'line-list',
+          'line-strip',
+          'triangle-list',
+          'triangle-strip',
+        ])
+      )
+  )
+  .fn(t => {
+    const { indexFormat, primitiveTopology } = t.params;
+
+    let indexArray: Uint16Array | Uint32Array;
+    if (indexFormat === 'uint16') {
+      if (primitiveTopology === 'triangle-list') {
+        indexArray = new Uint16Array([0, 1, 3, 0xffff, 2, 1, 0, 0]);
+      } else if (primitiveTopology === 'triangle-strip') {
+        indexArray = new Uint16Array([0, 1, 3, 0xffff, 2, 2]);
+      } else {
+        indexArray = new Uint16Array([0, 1, 0xffff, 2, 3, 3]);
+      }
+    } else {
+      if (primitiveTopology === 'triangle-list') {
+        indexArray = new Uint32Array([0, 1, 3, 0xffffffff, 2, 1, 0]);
+      } else if (primitiveTopology === 'triangle-strip') {
+        indexArray = new Uint32Array([0, 1, 3, 0xffffffff, 2]);
```

Comment:
If there were a bug where 'triangle-strip' were translated as 'triangle-list', I think this would still pass, because it would only be one triangle (0,1,3 followed by the incomplete triangle 0xffffffff,2).

Reply:
I think even without primitive restart, it's also pass for such bug. But here，I should draw another bottom right triangle to make the primitives after the restart values also can be renderred.